### PR TITLE
Adds the Vibebot

### PR DESCRIPTION
**Pull request by TheVekter**
## About The Pull Request

Adds the Vibebot. It's just a bot that vibes. Basically a disco ball on wheels. Craftable with:

1x Cyborg head
1x Crayon
2x Lightbulbs
1x Proximity Sensor

Needs to be made with the crafting system, there's no alternate construction using an item like the other bots.

## Why It's Good For The Game

Why isn't it? I don't know, the R&D rework has my brain fried and I needed something cute.

## Changelog
:cl:
add: Added the Vibebot
/:cl:

### CODE REVIEW

**Comment by Tharcoonvagh** [comment]
I'm just vibin' 'ere

**Comment by Anonmare** [comment]
Vibe check

**Comment by Buggy123** [comment]
This PR's got some great vibes.

**Comment by mrhugo13** [comment]
bro he vibin

**Comment by PKPenguin321** [comment]
I can't be mad, he's just vibing

**Comment by trollbreeder** [comment]
oh lawd he do be vibin tho

**Comment by PKPenguin321** [comment]
emag act that makes it shoot mindbreaker gas to give you bad vibes

**Comment by ATH1909** [comment]
> emag act that makes it shoot mindbreaker gas to give you bad vibes

*emag_act() that makes it blare PK borg harm alarms

**Comment by spessbandit** [comment]
adding vibrator bots wow

**Comment by gitbirb** [comment]
> > emag act that makes it shoot mindbreaker gas to give you bad vibes
> 
> *emag_act() that makes it blare PK borg harm alarms
better yet, just make it vibecheck everyone by giving it PK borg alarms (with vibe check instead of human harm) and syndicate revolver that it tries to fire occasionally

**Comment by zxaber** [comment]
>KNOWN ISSUES:
The whole thing changes colors instead of just the head like on the disco ball. I don't know how to fix this, the disco ball does it by having the head be something separate but I can't work out how to do that on a mob.

Couldn't you just use an overlay for that?

File: code/modules/mob/living/simple_animal/bot/vibebot.dm [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+	if(TimerID)
+		deltimer(TimerID)
+
+/mob/living/simple_animal/bot/vibebot/proc/Vibe()
+	remove_atom_colour(TEMPORARY_COLOUR_PRIORITY)
+	current_color = random_color()
+	set_light(range, power, current_color)
+	add_atom_colour("#[current_color]", FIXED_COLOUR_PRIORITY)
+	update_icon()
+	TimerID = addtimer(CALLBACK(src, .proc/Vibe), 5, TIMER_STOPPABLE)  //Call ourselves every 0.5 seconds to change colors
```
Comment (nemvar) [inline code/modules/mob/living/simple_animal/bot/vibebot.dm]: You should probably just add yourself to fastprocess instead.

Comment (TheVekter) [reply code/modules/mob/living/simple_animal/bot/vibebot.dm]: Not too sure how that works. I'll look into it tonight.

File: code/datums/components/crafting/recipes.dm [hunk 2]
```diff
@@ -217,6 +217,16 @@
 	time = 40
 	category = CAT_ROBOT
 
+/datum/crafting_recipe/Vibebot
+	name = "Vibebot"
+	result = /mob/living/simple_animal/bot/vibebot
+	reqs = list(/obj/item/light/bulb = 2,
+				/obj/item/bodypart/head/robot = 1,
+				/obj/item/assembly/prox_sensor = 1,
+				/obj/item/toy/crayon = 1)
```
Comment (ATH1909) [inline code/datums/components/crafting/recipes.dm]: no robot arm needed?

Comment (TheVekter) [reply code/datums/components/crafting/recipes.dm]: Nope, it uses a robot head instead. If you look at the sprite, there's a head in there (which I know is supposed to be from an Ethereal shut up)

File: code/modules/mob/living/simple_animal/bot/vibebot.dm [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+	. = ..()
+	update_icon()
+	auto_patrol = TRUE
+	Vibe()
+
+/mob/living/simple_animal/bot/vibebot/get_controls(mob/user)
+	var/dat
+	dat += hack(user)
+	dat += showpai(user)
+	dat += text({"
```
Comment (AnturK) [inline code/modules/mob/living/simple_animal/bot/vibebot.dm]: Using text() in anno domini 2020. You're also adding these directly instead of using Join.

File: code/modules/mob/living/simple_animal/bot/vibebot.dm [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+"<A href='?src=[REF(src)];power=[TRUE]'>[on ? "On" : "Off"]</A>" )
+
+	if(!locked || issilicon(user) || IsAdminGhost(user))
+		dat += text({"<BR> Auto Patrol: []"},
+
+"<A href='?src=[REF(src)];operation=patrol'>[auto_patrol ? "On" : "Off"]</A>" )
+	return	dat
+/mob/living/simple_animal/bot/vibebot/turn_on()
+	. = ..()
+	TurnedOn = TRUE //Mood
```
Comment (AnturK) [inline code/modules/mob/living/simple_animal/bot/vibebot.dm]: Bots already have default on variable why is this necessary.

File: code/modules/mob/living/simple_animal/bot/vibebot.dm (hunk 1 above)
Comment (AnturK) [reply code/modules/mob/living/simple_animal/bot/vibebot.dm]: Yes this should be on processing if it's constant.

File: code/modules/mob/living/simple_animal/bot/vibebot.dm [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+/mob/living/simple_animal/bot/vibebot/proc/Vibe()
+	remove_atom_colour(TEMPORARY_COLOUR_PRIORITY)
+	current_color = random_color()
+	set_light(range, power, current_color)
+	add_atom_colour("#[current_color]", FIXED_COLOUR_PRIORITY)
+	update_icon()
+	TimerID = addtimer(CALLBACK(src, .proc/Vibe), 5, TIMER_STOPPABLE)  //Call ourselves every 0.5 seconds to change colors
+
+/mob/living/simple_animal/bot/vibebot/proc/retaliate(mob/living/carbon/human/H)
```
Comment (AnturK) [inline code/modules/mob/living/simple_animal/bot/vibebot.dm]: If this is intended override add return.

File: code/modules/mob/living/simple_animal/bot/vibebot.dm [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+	var/TurnedOn = FALSE
+	var/current_color
+	var/TimerID
+	var/range = 7
+	var/power = 3
+
+/mob/living/simple_animal/bot/vibebot/Initialize()
+	. = ..()
+	update_icon()
+	auto_patrol = TRUE
```
Comment (AnturK) [inline code/modules/mob/living/simple_animal/bot/vibebot.dm]: You can just make this default.

**Review by AnturK — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/simple_animal/bot/vibebot.dm [hunk 7]
```diff
@@ -18,17 +18,15 @@
...
-	var/TimerID
 	var/range = 7
 	var/power = 3
 
 /mob/living/simple_animal/bot/vibebot/Initialize()
 	. = ..()
 	update_icon()
 	auto_patrol = TRUE
-	Vibe()
+	on = TRUE
```
Comment (AnturK) [inline code/modules/mob/living/simple_animal/bot/vibebot.dm]: Pretty sure bots start on by default so no need for this.

File: code/modules/mob/living/simple_animal/bot/vibebot.dm [hunk 8]
```diff
@@ -49,27 +47,22 @@ Maintenance panel panel is [open ? "opened" : "closed"]"},
 	return	dat
 /mob/living/simple_animal/bot/vibebot/turn_on()
 	. = ..()
-	TurnedOn = TRUE //Mood
+	on = TRUE
```
Comment (AnturK) [inline code/modules/mob/living/simple_animal/bot/vibebot.dm]: Same here, this and update_icon will be done in parent call already.

File: code/modules/mob/living/simple_animal/bot/vibebot.dm [hunk 9]
```diff
@@ -49,27 +47,22 @@ Maintenance panel panel is [open ? "opened" : "closed"]"},
...
 	update_icon()
 	Vibe()
 
 /mob/living/simple_animal/bot/vibebot/turn_off()
 	. = ..()
-	TurnedOn = FALSE
-	update_icon()
-	set_light(0)
+	on = FALSE
```
Comment (AnturK) [inline code/modules/mob/living/simple_animal/bot/vibebot.dm]: Already set in parent

File: code/modules/mob/living/simple_animal/bot/vibebot.dm [hunk 10]
```diff
@@ -86,5 +79,8 @@ Maintenance panel panel is [open ? "opened" : "closed"]"},
 		if(mode == BOT_PATROL)
 			bot_patrol()
 
-///obj/structure/vibebot/update_icon_state()
-//	icon_state = "vibebot_head_[TurnedOn]"
+	if(on == TRUE)
```
Comment (AnturK) [inline code/modules/mob/living/simple_animal/bot/vibebot.dm]: just `if(on)` is enough

File: code/modules/mob/living/simple_animal/bot/vibebot.dm [hunk 11]
```diff
@@ -86,5 +79,8 @@ Maintenance panel panel is [open ? "opened" : "closed"]"},
...
+		Vibe()
+
+	if(on == FALSE)
```
Comment (AnturK) [inline code/modules/mob/living/simple_animal/bot/vibebot.dm]: just `else` here.

**Review by AnturK — CHANGES_REQUESTED** [review]

**Comment by BadSS13Player** [comment]
Yeah, you need to make an emagged vibe bot do a vibe check.

**Comment by TheSilverNuke** [comment]
What does this do when emagged by the way?

**Comment by TheVekter** [comment]
Nothing, as far as I've coded. I need to add an emag_act but the code for it is kinda spaghetti and I haven't had much free time lately.
